storage: Return an error when a non-pending intent is found

kkaneda opened the pull request:
Address a TODO comment in intentResolver.maybePushTransactions.

Confirmed that
- all intents created in mvcc.go have the PENDING status, and
- all tests passed even if we call panic() when a non-pending intent is found.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5470)

<!-- Reviewable:end -->

kkaneda committed:
storage: Return an error when a non-pending intent is found

Address a TODO comment in intentResolver.maybePushTransactions.

Confirmed that
- all intents created in mvcc.go have the PENDING status, and
- all tests passed even if we call panic() when a non-pending intent is found.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/intent_resolver.go, line 176 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470#-KDTsfRViBtCq3OtBk7x:-KDTsfRViBtCq3OtBk7y:506780002) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a191b8e7d70a4855558265f6ae4d2c2cf5bebb2/storage/intent_resolver.go#L176)):*
maybe leave this as before (since the `resolveIntents` slice will end up being created anyway), and then return this error after this loop. That way, we'd get all the "offending" intents, rather than just the first one.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470#-:-KDTszmlfcdGQlGkTpjK:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on storage/intent_resolver.go:
```diff
@@ -220,11 +224,11 @@ func (ir *intentResolver) maybePushTransactions(ctx context.Context, intents []r
 	}
 	ir.mu.Unlock()
-	if err != nil {
-		// TODO(bdarnell): return resolveIntents even on error.
```
why'd this TODO go?

bdarnell commented:
LGTM

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/intent_resolver.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470#-KDUJCNLUoEjbD9daUsw:-KDUJCNLUoEjbD9daUsx:1056022594) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a191b8e7d70a4855558265f6ae4d2c2cf5bebb2/storage/intent_resolver.go#L130)):*
Delete the rest of this comment starting with "regardless".

At each of the two places we call `maybePushTransactions` move the call to `resolveIntents` until after we've handled the error from `maybePushTransactions`.

---

*[storage/intent_resolver.go, line 176 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470#-KDTsfRViBtCq3OtBk7x:-KDUJOeNNOioSTqEK0gc:-1411890176) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a191b8e7d70a4855558265f6ae4d2c2cf5bebb2/storage/intent_resolver.go#L176)):*
I think it's fine as is; this is an assertion for something that should never happen.

---

*[storage/intent_resolver.go, line 224 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470#-KDUK9NczOKiFs-ixhU0:-KDUK9NczOKiFs-ixhU1:580942878) ([raw file](https://github.com/cockroachdb/cockroach/blob/67899cd92375bf62cffd8a116a8ad3ef910406eb/storage/intent_resolver.go#L224)):*
@tschottdorf This comment is gone because there are no longer any `resolveIntents` to return at this point.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470#-:-KDUKH5D9D7AZeK4nEHK:676834233)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/intent_resolver.go, line 176 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470#-KDTsfRViBtCq3OtBk7x:-KDULqGhipYgy0BEhgrY:-1000574931) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a191b8e7d70a4855558265f6ae4d2c2cf5bebb2/storage/intent_resolver.go#L176)):*
Maybe it's fine, but we are getting into a habit of emitting less-useful-than-is-possible error messages, and I'd like to see that habit broken.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470)*
<!-- Sent from Reviewable.io -->

kkaneda commented:
Thanks for the reviews! 

---

Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/intent_resolver.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470#-KDUJCNLUoEjbD9daUsw:-KDW4TMkCYNNjDZqh-bF:171199004) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a191b8e7d70a4855558265f6ae4d2c2cf5bebb2/storage/intent_resolver.go#L130)):*
Ah, I see. Removed.

---

*[storage/intent_resolver.go, line 176 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470#-KDTsfRViBtCq3OtBk7x:-KDW66J6hG6T0rJCfsFo:-6366582) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a191b8e7d70a4855558265f6ae4d2c2cf5bebb2/storage/intent_resolver.go#L176)):*
I see. Sure, changed to collect all non-pending intents and return an error after the loop. Renamed `resolveIntents` to `nonPendingIntents`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470#-:-KDWD_U7CMMKp3JsjKmp:1382943234)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5470#-:-KDX8AzqLT6Ym8QWjbeU:1417088721)*
<!-- Sent from Reviewable.io -->
